pkg/gameserverallocations/find.go

```diff
@@ -40,6 +40,7 @@ func findGameServerForAllocation(gsa *allocationv1.GameServerAllocation, list []
 	var loop func(list []*agonesv1.GameServer, f func(i int, gs *agonesv1.GameServer))
 
 	// packed is forward looping, distributed is random looping
+	// TODO: Discuss if we change findGameServerForAllocation so all strategies use forward looping for the the pre-sorted list for each strategy once FeatureCountsAndLists gate is graduated.
```

Comment:
For the new version, I think we can throw most of this complexity away, and do simple search and filter -- we don't need this fancy looping construct anymore, since it will be pre-sorted / pre-randomised.

(Also can then also do a duplicated function as well for counts and lists)

Reply:
Right now we have three paths A) Sort by Allocated and then loop through list. B) Sort by Allocated, then randomly shuffle list, and then loop through randomized list. C) Sort by CountsAndLists and then loop through list. 

Where/when should I be randomizing? Should I make a 4th path D) Sort by CountsAndLists, then randomly shuffle list, and then loop through randomized list? Or are you suggesting a different set of paths?

Reply:
My thought is this - for this part of the code -- the list should already be appropriately sorted when it is passed to it. So it should be a `for .. range` operation.

Then for each strategy:

_Distributed_
1. Randomise first (so that will give us a random distribution within each sorted Allocated/Ready section of the list we have)
2. Sort by:
  3. Allocated
  4. Count/List as appropriate

Question on the above: Does K8s List already give us a generally random order of GameServers, such that we don't need to bother shuffling it? I'm actually not sure! May be worth checking.

_Packed_
Sort by:
1. Allocated First
2. Packed Strategy (basically what we have now)
3. Tie break with counts and lists

I hope that makes things clearer?

Reply:
Why randomize before sorting? I thought we had the entire list of Ready/Allocated game servers here, and not a subset?

If we're keeping our current definitions of Packed and Distributed, then it seems to make more sense to more or less keep the flow as-is:
1.0 Sort by Allocated
1.1 Tie break sort by number of Ready servers on the Node
1.2 Tie break sort by Counts and Lists
1.3 Tie break sort by Game Server Name <- Separate discussion on this as it currently tie breaks on NodeName
2.0 If Distributed, randomize list
2.1 Loop through list to find matching Game Servers (original sorted list for Packed, randomized list for Distributed) 

A possible alternative would to first check for matches before sorting the Game Servers, and only sort for the Packed case:
1. Filter List of Game Servers for Matches
2.a Return unsorted list of Matches for Distributed (completely random, no sorting) <- I'd need to ask an expert in Go if this is dependably random, as this depends on how c.getGameServers() ranges through the cache map of Game Servers
2.b Sort List (same as above) and return sorted list of Matches for Packed

Reply:
With the above being said, I think the current `findGameServerForAllocation` totally works fine. We can leave all the refactoring outside this PR.